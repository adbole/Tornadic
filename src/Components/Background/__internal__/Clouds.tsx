import React from "react";
import { Cloud } from "@react-three/drei";

import type WeatherCondition from "ts/WeatherCondition";


export default function Clouds({ isDay, condition }: { isDay: boolean, condition: WeatherCondition["type"] }) {
    const [show, setShow] = React.useState(false)

    React.useEffect(() => {
        switch(condition) {
            case "Clear":
                setShow(false)
                break
            default:
                setShow(true)
        }
    }, [condition])

    const cloudsType = React.useMemo(() => {
        switch(condition) {
            case "Overcast":
            case "Rain":
            case "Snow":
            case "Snow Grains":
            case "Thunderstorms":
            case "Thunderstorms and Hail":
                return "Heavy"
            case "Rain Showers":
            case "Snow Showers":
            case "Partly Cloudy":
                return "Medium"
            default:
                return "Light"
        }
    }, [condition])

    const segments = React.useMemo(() => {
        if(cloudsType === "Heavy") return 200;
        else if(cloudsType === "Medium") return 100;

        return 50;
    }, [cloudsType]);


    const opacity = React.useMemo(() => {
        if(cloudsType === "Heavy") return 1;
        else if(cloudsType === "Medium") return 0.4;

        return 0.25;
    }, [cloudsType]);

    return (
        <group visible={show}>
            <ambientLight intensity={isDay ? 4 : 0.2} />
            <Cloud 
                seed={1}
                position={[0, 10, -10]}
                segments={segments}
                opacity={opacity}
                growth={5}
                speed={0.2}
                bounds={[30, 1, 1]}
            />
        </group>
    )
}