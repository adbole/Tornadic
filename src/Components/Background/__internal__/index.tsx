export { default as Clouds } from "./Clouds";
export { default as RainSnow } from "./RainSnow";
export { default as Thunder } from "./Thunder";