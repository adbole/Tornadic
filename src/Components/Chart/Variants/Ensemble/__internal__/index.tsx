export * from "./Constants";
export { default as Outliers } from "./Outliers";
export { default as TooltipDisplay } from "./TooltipDisplay";
