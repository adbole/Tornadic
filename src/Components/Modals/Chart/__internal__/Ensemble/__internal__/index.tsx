export { default as Outliers } from "./Outliers";
export { default as TooltipDisplay } from "./TooltipDisplay";