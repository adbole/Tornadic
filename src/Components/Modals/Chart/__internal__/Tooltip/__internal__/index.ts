export { default as PrimaryInformation } from "./PrimaryInformation";
export { default as SecondaryInformation } from "./SecondaryInformation";
export { default as Time } from "./Time";
