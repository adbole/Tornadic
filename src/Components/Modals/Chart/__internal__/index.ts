export { default as ChartDisplay } from "./ChartDisplay"
export { default as CustomTooltip } from "./CustomTooltip"
export { default as getDataVisual } from "./getDataVisual"