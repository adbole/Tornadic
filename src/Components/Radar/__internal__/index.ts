export { default as AlertPolygons } from "./AlertPolygons"
export { Position } from "./ControlPortal"
export { default as ControlPortal } from "./ControlPortal"
export { default as Home } from "./Home"
export { default as Locate } from "./Locate"
export { default as Opacity } from "./Opacity"
export { default as Peek } from "./Peek"
export { default as PlayPauseButton } from "./PlayPauseButton"
export { default as RainViewer } from "./RainViewer"