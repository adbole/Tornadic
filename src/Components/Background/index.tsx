import React from "react";
import { Global } from "@emotion/react";
import styled from "@emotion/styled";
import { Sky } from "@react-three/drei";
import { Canvas } from "@react-three/fiber";

import { useReadLocalStorage } from "Hooks";

import { useWeather } from "Contexts/WeatherContext";

import { Clouds, Gradient, RainSnow, Stars, Thunder } from "./__internal__";


const StyledCanvas = styled(Canvas)({
    top: 0,
    left: 0,
    zIndex: -1,
});

function HighConstrast({ rayleigh }: { rayleigh: number }) {
    const settings = useReadLocalStorage("userSettings");

    if (rayleigh !== 1 || !settings?.highContrastForLive) return null;

    return <Global styles={{ "#root": { color: "#3f3f3f" } }} />;
}

export default function Background() {
    const { weather } = useWeather();
    const settings = useReadLocalStorage("userSettings");

    const { isDay, condition } = React.useMemo(
        () => ({
            isDay: weather.isDay(),
            condition: weather.getNow().conditionInfo.type,
        }),
        [weather]
    );

    const [rayleigh, elevation, exposure] = React.useMemo(() => {
        if (isDay) {
            switch (condition) {
                case "Clear":
                case "Mostly Clear":
                    return [1, 0.75, 0.25];
                default:
                    return [5, 0.75, 0.1];
            }
        }

        return [0.01, 0.75, 0.2];
    }, [isDay, condition]);

    return (
        <>
            <HighConstrast rayleigh={rayleigh} />
            <Gradient isDay={isDay} condition={condition} />
            
            {settings && !settings.preferGradient && (
                <StyledCanvas
                    style={{ position: "fixed" }}
                    gl={{ toneMappingExposure: exposure }}
                >
                    <Sky
                        rayleigh={rayleigh}
                        inclination={elevation}
                        mieCoefficient={0.005}
                        mieDirectionalG={0.99}
                        turbidity={1}
                        azimuth={0.25}
                    />
                    <Clouds isDay={isDay} condition={condition} />
                    <RainSnow condition={condition} />
                    <Thunder condition={condition} />
                    <Stars isDay={isDay} condition={condition} />
                </StyledCanvas>
            )}
        </>
    );
}
