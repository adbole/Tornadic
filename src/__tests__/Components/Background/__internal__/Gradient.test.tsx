import { render } from "@testing-library/react";

import { Gradient } from "Components/Background/__internal__";

import type WeatherCondition from "ts/WeatherCondition";


const mocks = vi.hoisted(() => ({
    Global: vi.fn(),
}))

vi.mock("@emotion/react", async (importOriginal) => ({
    ...(await importOriginal() as any),
    Global: mocks.Global,
}))

describe.each([
    ["Day", true],
    ["Night", false],
])("%s", (_, isDay) => {
    test.each([
        "Clear",
        "Drizzle",
        "Foggy",
        "Freezing Drizzle",
        "Freezing Rain",
        "Mostly Clear",
        "Overcast",
        "Partly Cloudy",
        "Rain",
        "Rain Showers",
        "Snow",
        "Snow Grains",
        "Snow Showers",
        "Thunderstorms",
        "Thunderstorms and Hail"
    ] as WeatherCondition["type"][])("%s", (condition) => {
        render(
            <Gradient isDay={isDay} condition={condition} />
        )

        expect(mocks.Global.mock.lastCall[0].styles.styles).toMatchSnapshot()
    })
})