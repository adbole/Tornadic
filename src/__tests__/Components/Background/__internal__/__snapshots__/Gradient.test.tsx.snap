// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Day > Clear 1`] = `"body{background:linear-gradient(to bottom, #7a98b2, #bfc5c6);};label:Gradient;"`;

exports[`Day > Drizzle 1`] = `"body{background:linear-gradient(to bottom, #0f1c50, #8da3bd);};label:Gradient;"`;

exports[`Day > Foggy 1`] = `"body{background:linear-gradient(to bottom, #7a98b2, #bfc5c6);};label:Gradient;"`;

exports[`Day > Freezing Drizzle 1`] = `"body{background:linear-gradient(to bottom, #0f1c50, #8da3bd);};label:Gradient;"`;

exports[`Day > Freezing Rain 1`] = `"body{background:linear-gradient(to bottom, #0f1c50, #8da3bd);};label:Gradient;"`;

exports[`Day > Mostly Clear 1`] = `"body{background:linear-gradient(to bottom, #7a98b2, #bfc5c6);};label:Gradient;"`;

exports[`Day > Overcast 1`] = `"body{background:linear-gradient(to bottom, #acb7bd, #73bae1);};label:Gradient;"`;

exports[`Day > Partly Cloudy 1`] = `"body{background:linear-gradient(to bottom, #7a98b2, #bfc5c6);};label:Gradient;"`;

exports[`Day > Rain 1`] = `"body{background:linear-gradient(to bottom, #0f1c50, #8da3bd);};label:Gradient;"`;

exports[`Day > Rain Showers 1`] = `"body{background:linear-gradient(to bottom, #0f1c50, #8da3bd);};label:Gradient;"`;

exports[`Day > Snow 1`] = `"body{background:linear-gradient(to bottom, #797c87, #acb7bd);};label:Gradient;"`;

exports[`Day > Snow Grains 1`] = `"body{background:linear-gradient(to bottom, #797c87, #acb7bd);};label:Gradient;"`;

exports[`Day > Snow Showers 1`] = `"body{background:linear-gradient(to bottom, #797c87, #acb7bd);};label:Gradient;"`;

exports[`Day > Thunderstorms 1`] = `"body{background:linear-gradient(to bottom, #4f5158, #8da3bd);};label:Gradient;"`;

exports[`Day > Thunderstorms and Hail 1`] = `"body{background:linear-gradient(to bottom, #4f5158, #8da3bd);};label:Gradient;"`;

exports[`Night > Clear 1`] = `"body{background:linear-gradient(to bottom, #1a1b1c, #1a1b1c);};label:Gradient;"`;

exports[`Night > Drizzle 1`] = `"body{background:linear-gradient(to bottom, #0f1c50, #8da3bd);};label:Gradient;"`;

exports[`Night > Foggy 1`] = `"body{background:linear-gradient(to bottom, #1a1b1c, #1a1b1c);};label:Gradient;"`;

exports[`Night > Freezing Drizzle 1`] = `"body{background:linear-gradient(to bottom, #0f1c50, #8da3bd);};label:Gradient;"`;

exports[`Night > Freezing Rain 1`] = `"body{background:linear-gradient(to bottom, #0f1c50, #8da3bd);};label:Gradient;"`;

exports[`Night > Mostly Clear 1`] = `"body{background:linear-gradient(to bottom, #1a1b1c, #1a1b1c);};label:Gradient;"`;

exports[`Night > Overcast 1`] = `"body{background:linear-gradient(to bottom, #2d3438, #1a1b1c);};label:Gradient;"`;

exports[`Night > Partly Cloudy 1`] = `"body{background:linear-gradient(to bottom, #1a1b1c, #1a1b1c);};label:Gradient;"`;

exports[`Night > Rain 1`] = `"body{background:linear-gradient(to bottom, #0f1c50, #8da3bd);};label:Gradient;"`;

exports[`Night > Rain Showers 1`] = `"body{background:linear-gradient(to bottom, #0f1c50, #8da3bd);};label:Gradient;"`;

exports[`Night > Snow 1`] = `"body{background:linear-gradient(to bottom, #797c87, #acb7bd);};label:Gradient;"`;

exports[`Night > Snow Grains 1`] = `"body{background:linear-gradient(to bottom, #797c87, #acb7bd);};label:Gradient;"`;

exports[`Night > Snow Showers 1`] = `"body{background:linear-gradient(to bottom, #797c87, #acb7bd);};label:Gradient;"`;

exports[`Night > Thunderstorms 1`] = `"body{background:linear-gradient(to bottom, #4f5158, #8da3bd);};label:Gradient;"`;

exports[`Night > Thunderstorms and Hail 1`] = `"body{background:linear-gradient(to bottom, #4f5158, #8da3bd);};label:Gradient;"`;
